aspnet/fundamentals/routing.rst

```diff
@@ -1,10 +1,283 @@
+
+The example below shows how to generate a link to a route given a dictionary of route values and a ``RouteCollection``.
+
+.. literalinclude:: routing/sample/RoutingSample/Startup.cs
+  :language: c#
+  :lines: 45-69
+  :dedent: 8
+  :emphasize-lines: 9,14,17-18,23
+
+The ``VirtualPath`` generated at the end of the sample above is ``/package/create/123``.
+
+The second parameter to the ``VirtualPathContext`` constructor is a collection of `ambient values`. Ambient values provide convenience by limiting the number of values a developer must specify within a certain request context. The current route values of the current request are considered ambient values for link generation. For example, in an MVC application if you are in the About action of the HomeController, you don't need to specify the controller route value to link to the Index action (the ambient value of Home will be used). 
+
+Ambient values that don't match a parameter are ignored, and ambient values are also ignored when an explicitly-provided value overrides it, going from left to right in the URL.
+
+Values that are explicitly provided but which don't match anything are added to the query string.
+
+.. list-table:: Generating Links
+	:header-rows: 1
+
+	* - Matched Route
+	  - Ambient Values
+	  - Explicit Values
+	  - Result
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home"
+	  - action="About"
+	  - ``/Home/About``
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home"
+	  - controller="Order",action="About"
+	  - ``/Order/About``
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home",color="Red"
+	  - action="About"
+	  - ``/Home/About``
+	* - ``{controller}/{action}/{id?}``
+	  - controller="Home"
+	  - action="About",color="Red"
+	  - ``/Home/About?color=Red``
+
+If a route parameter has a default value that doesn't match a parameter and that value is explicitly provided, it must match the default value. For example:
+
+.. code-block:: c#
+	
+	routes.MapRoute("blog_route", "blog/{*slug}", 
+		defaults: new { controller = "Blog", action = "ReadPost" });
+	
+Link generation would only generate a link for this route when the matching values for controller and action are provided.
+
+Recommendations
+---------------
+
+Routing is a powerful feature that is built into the default ASP.NET MVC project template such that most apps will be able to leverage it without having to customize its behavior. This is by design; customizing routing behavior is an advanced development approach. Keep in mind the following recommendations with regard to routing:
+
+	- Most apps shouldn't need custom routes. The default route will work in most cases.
+	- Don't use custom routes unless you understand them well and are sure you need them.
+	- Routes can be tricky to test and debug.
+	- Routes should not be used as a means of securing your controllers or their action methods.
```

Comment:
Noted. Not sure where we have anything separate in the docs for attribute routing.